File: ruby/lib/TituloPropiedad.rb
module Civitas

    class TituloPropiedad
        
        private
        @@factorInteresesHipoteca=1.1

        def esEsteElPropietario(jugador)
            
        end

        def getImporteHipoteca

        end

        def getPrecioAlquiler
            if (!@hipotecado || propietarioEncarcelado)
                return (@alquilerBase*(1+(@numCasas*0.5)+(numHoteles*2.5)))
            else return 0
        end

        def getPrecioVenta
            
        end

        def propietarioEncarcelado

        end
        
        public
        def initialize(nom, ab, fr, hb, pc, pe)
            @alquilerBase=ab
            @factorRevalorizacion=fr
            @hipotecaBase=hb
            @hipotecado=false
            @nombre=nom
            @numCasas=0
            @numHoteles=0
            @precioCompra=pc
            @precioEdificar=pe
        end

        def toString
            string="Alquiler Base= "+@alquilerBase
            +", Factor Revalorización: "+@factorRevalorizacion
            +", Hipoteca Base: "+@hipotecaBase
            +", Hipotecado: "+@hipotecado
            +", Nombre: "+@nombre
            +", Casas: "+@numCasas
            +", Hoteles: "+@numHoteles
            +", Precio compra: "+@precioCompra
            +", Precio casa: "+@precioEdificar

            return string
        end

        def getImporteCancelarHipoteca
            return (@@factorInteresesHipoteca*@hipotecaBase)
        end

        def cancelarHipoteca (jugador)
            if (@hipotecado && esEsteElPropietario(jugador))
                jugador.paga(getImporteCancelarHipoteca)
                @hipotecado=false
                return true
            else
                return false
            end
        end

        def actualizaPropietarioPorConversion(jugador)
            
        end

        def cantidadCasasHoteles
            
        end

        def comprar(jugador)
            
        end

        def construirCasa(jugador)
            
        end

        def construirHotel(jugador)
            
        end

        def derruirCasas(n, jugador)
            
        end

        def hipotecar(jugador)
            if (!@hipotecado && )
                
            else
                
            end
        end

        def tienePropietario(jugador)

        end

        def tramitarAlquiler(jugador)

        end

        def vender(jugador)
        
        end
    end
end

File: java/src/civitas/TituloPropiedad.java
package civitas;

public class TituloPropiedad {

    private float alquilerBase;
    private static float factorInteresesHipoteca = (float) 1.1;
    private float factorRevalorizacion;
    private float hipotecaBase;
    private boolean hipotecado;
    private String nombre;
    private int numCasas;
    private int numHoteles;
    private float precioCompra;
    private float precioEdificar;
    private Jugador propietario;

    TituloPropiedad(String nom, float ab, float fr, float hb, float pc, float pe) {
        alquilerBase = ab;
        factorRevalorizacion = fr;
        hipotecaBase = hb;
        hipotecado = false;
        nombre = nom;
        numCasas = 0;
        numHoteles = 0;
        precioCompra = pc;
        precioEdificar = pe;
        propietario = null;
    }

    public String toString() {
        return "Nombre: " + nombre + "\nAlquiler base: " + alquilerBase + "\nFactor de Revalorización: "
                + factorRevalorizacion + "\nHipoteca Base: " + hipotecaBase + "\nHipotecado: " + hipotecado
                + "\nNúmero de CASAS: " + numCasas + "\nNúmero de HOTELES: " + numHoteles + "\nPrecio de COMPRA: "
                + precioCompra + "\nPrecio para EDIFICAR: " + precioEdificar + "\nPropietario: "
                + propietario.toString();
    }

    private float getPrecioAlquiler() {
        if (hipotecado || propietarioEncarcelado()) {
            return 0;
        }

        else {
            return (float) (alquilerBase * (1 + (numCasas * 0.5) + (numHoteles * 2.5)));
        }
    }

    private float getImporteHipoteca() {
        return hipotecaBase * (1 + (numCasas * 0.5f) + (numHoteles * 2.5f));
    }

    float getImporteCancelarHipoteca() {
        return getImporteHipoteca() * factorInteresesHipoteca;
    }

    boolean cancelarHipoteca(Jugador jugador) {
        if (hipotecado && esEsteElPropietario(jugador)) {
            jugador.paga(getImporteCancelarHipoteca());
            hipotecado = false;
            return true;
        }

        else {
            return false;
        }
    }

    boolean hipotecar(Jugador jugador) {
        if (!hipotecado && esEsteElPropietario(jugador)) {
            jugador.recibe(getImporteHipoteca());
            hipotecado = true;
            return true;
        } else {
            return false;
        }
    }

    void tramitarAlquiler(Jugador jugador) {
        if (tienePropietario() && !esEsteElPropietario(jugador)) {
            float precio = getPrecioAlquiler();
            jugador.pagaAlquiler(precio);
            getPropietario().recibe(precio);
        }
    }

    boolean propietarioEncarcelado() {
        return getPropietario().isEncarcelado();
    }

    int cantidadCasasHoteles() {
        return numCasas + numHoteles;
    }

    boolean destruirCasas(int n, Jugador jugador) {
        if (esEsteElPropietario(jugador) && numCasas >= n) {
            numCasas -= n;
            return true;
        }
        return false;
    }

    float getPrecioVenta() {
        return precioCompra + (numCasas + 5 * numHoteles) * (precioEdificar * factorRevalorizacion);
    }

    boolean construirCasa(Jugador jugador) {
        if (esEsteElPropietario(jugador)) {
            jugador.paga(precioEdificar);
            numCasas++;
            return true;
        }

        else {
            return false;
        }
    }

    boolean construirHotel(Jugador jugador) {
        if (esEsteElPropietario(jugador)) {
            jugador.paga(precioEdificar);
            numHoteles++;
            return true;
        }

        else {
            return false;
        }

    }

    boolean comprar(Jugador jugador) {
        if (tienePropietario()) {
            return false;
        } else {
            propietario = jugador;
            jugador.paga(precioCompra);
            return true;
        }
    }

    private boolean esEsteElPropietario(Jugador jugador) {
        return (jugador.getNombre() == propietario.getNombre());
    }

    boolean tienePropietario() {
        return propietario != null;
    }

    Jugador getPropietario() {
        return propietario;
    }

    int getNumCasas() {
        return numCasas;
    }

    String getNombre() {
        return nombre;
    }

    public boolean getHipotecado() {
        return hipotecado;
    }

    int getNumHoteles() {
        return numHoteles;
    }

    float getPrecioCompra() {
        return precioCompra;
    }

    float getPrecioEdificar() {
        return precioEdificar;
    }

    boolean vender(Jugador jugador) {
        if (esEsteElPropietario(jugador)) {
            propietario = null;
            jugador.recibe(getPrecioVenta());
            numCasas = 0;
            numHoteles = 0;
            return true;
        }
        return false;
    }

    void actualizaPropietarioPorConversion(Jugador jugador) {
        if (!esEsteElPropietario(jugador))
            propietario = jugador;
    }

}
